// ng2-animate-on-scroll/esm5/lib/animate-on-scroll.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AnimateOnScrollDirective } from './animate-on-scroll.directive';
import { ScrollService } from './scroll.service';
var AnimateOnScrollModule = /** @class */ (function () {
    function AnimateOnScrollModule() {
    }
    AnimateOnScrollModule_1 = AnimateOnScrollModule;
    AnimateOnScrollModule.forRoot = function () {
        return {
            ngModule: AnimateOnScrollModule_1,
            providers: [ScrollService]
        };
    };
    var AnimateOnScrollModule_1;
    AnimateOnScrollModule = AnimateOnScrollModule_1 = __decorate([
        NgModule({
            imports: [
                CommonModule
            ],
            declarations: [
                AnimateOnScrollDirective
            ],
            exports: [
                AnimateOnScrollDirective
            ]
        })
    ], AnimateOnScrollModule);
    return AnimateOnScrollModule;
}());
export { AnimateOnScrollModule };
