// ng2-animate-on-scroll/esm5/lib/scroll.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Subscription, empty, fromEvent } from 'rxjs';
var ScrollService = /** @class */ (function () {
    function ScrollService() {
        var _this = this;
        this.scrollSub = new Subscription();
        this.resizeSub = new Subscription();
        // set initial value
        this.manageScrollPos();
        // create observable that we can subscribe to from component or directive
        this.scrollObs = typeof window !== 'undefined' ? fromEvent(window, 'scroll') : empty();
        // initiate subscription to update values
        this.scrollSub = this.scrollObs
            .subscribe(function () { return _this.manageScrollPos(); });
        // create observable for changes in screen size
        this.resizeObs = typeof window !== 'undefined' ? fromEvent(window, 'resize') : empty();
        // initiate subscription to update values
        this.resizeSub = this.resizeObs
            .subscribe(function () { return _this.manageScrollPos(); });
    }
    ScrollService.prototype.manageScrollPos = function () {
        // update service property
        this.pos = typeof window !== 'undefined' ? window.pageYOffset : 0;
    };
    ScrollService.prototype.ngOnDestroy = function () {
        this.scrollSub.unsubscribe();
        this.resizeSub.unsubscribe();
    };
    ScrollService = __decorate([
        Injectable()
    ], ScrollService);
    return ScrollService;
}());
export { ScrollService };
